### DISCUSSION TITLE
Submodule mode / update

### DISCUSSION THREAD

**Discussion by m-u-xyz** [gitpython-developers/GitPython#1575] [Q&A]
Updating the submodule references in a supermodule's tree.
```
for so in repo.submodules:
    sn = git.objects.submodule.base.Submodule(r,rr.head.commit.binsha,name=so.name,path=so.path, mode=0o160000)
    repo.index.add([sn])
```
(A) an explicit "mode" is required. Given that anything that's not 0o160000 is very likely to break the index file this doesn't make much sense.

(B) There really should be a method that returns a new `Submodule` object, either using an explicit `binsha` or defaulting to the submodule's current HEAD, instead of forcing the user to create one from scratch.

**Comment by Byron** [comment]
[In the docs for submodules](https://gitpython.readthedocs.io/en/stable/tutorial.html#submodule-handling) there is suggestions for how the API is intended.

Generally, these are very buggy and I wouldn't use them for more than discovery. Further, the index API is superficial at best and again I don't recommend to use it especially when handling submodules. Instead, prefer `repo.git.add(…)`.

**Comment by m-u-xyz** [comment]
Well, I did try `repo.add(PATH)`, but that method has the unfortunate side effect of completely ignoring the fact that PATH is a submodule; it adds all the files in it to the supermodule's index instead.

The `repo.index.add()` method I quoted above at least works. It's even documented.
